src/BaGet.Decompiler/AssemblyDecompilerService.cs

```diff
@@ -0,0 +1,184 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using BaGet.Decompiler.Decompilation;
+using BaGet.Decompiler.Objects;
+using BaGet.Decompiler.SourceCode;
+using ICSharpCode.Decompiler;
+using ICSharpCode.Decompiler.CSharp;
+using ICSharpCode.Decompiler.CSharp.Syntax;
+using Mono.Cecil;
+
+namespace BaGet.Decompiler
+{
+    public class AssemblyDecompilerService
+    {
+        private readonly DecompilerSettings _decompilerMembersSettings;
+        private readonly Filter _filter;
+
+        private readonly List<ISourceCodeProvider> _sourceCodeProviders;
+
+        public AssemblyDecompilerService()
+        {
+            _decompilerMembersSettings = new DecompilerSettings
+            {
+                //DecompileMemberBodies = false,    TODO: https://github.com/icsharpcode/ILSpy/issues/1332
+                UsingDeclarations = false,
+                ShowXmlDocumentation = false
+            };
+
+            _filter = new Filter();
+
+            _sourceCodeProviders = new List<ISourceCodeProvider>
+            {
+                new SourceCodeDecompiler()
+            };
+        }
+
+        public AnalysisAssembly AnalyzeAssembly(Stream assembly, Stream pdb = null, Stream documentationXml = null)
+        {
+            string tempDir = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
+            Directory.CreateDirectory(tempDir);
+
+            AnalysisAssembly res;
+            try
+            {
+                string assemblyFile = Path.Combine(tempDir, "Binary.dll");
```

Comment:
You should be able to use [`StreamExtensions.AsTemporaryFileStreamAsync`](https://github.com/loic-sharma/BaGet/blob/master/src/BaGet.Core/Extensions/StreamExtensions.cs#L22) here.

Reply:
Maybe, although:

* I need the file name (can probably get that from the FileStream)
* I'm convinced the PDB file needs the same name, but different extension, as the dll file. Hasn't been doublechecked though.

Reply:
I've updated to ILSpy's 4.x packages, and it seems like they've written everything from scratch. It's entirely possible that we can do everything in memory now.

Reply:
On-disk files are no longer required. Yay.